File: algorithms/python/lastStoneWeightII/lastStoneWeightII.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/last-stone-weight-ii/
# Author: Miao Zhang
# Date:   2021-04-06

class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        dp = [0] * 1501
        sums = 0
        for st in stones:
            sums += st
        target = sums // 2
        for i in range(len(stones)):
            for j in range(target, stones[i] - 1, -1):
                dp[j] = max(dp[j], dp[j - stones[i]] + stones[i])
        return sums - dp[target] * 2

File: algorithms/cpp/lastStoneWeightII/lastStoneWeightII.cpp
// Source: https://leetcode.com/problems/last-stone-weight-ii/
// Author: Miao Zhang
// Date:   2021-04-06

class Solution {
public:
    int lastStoneWeightII(vector<int>& stones) {
        vector<bool> dp(1501);
        dp[0] = true;
        int sums = 0;
        for (int stone: stones) {
            sums += stone;
            for (int i = min(1500, sums); i >= stone; i--) {
                dp[i] = dp[i] | dp[i - stone];
            }
        }
        for (int i = sums / 2; i >= 0; i--) {
            if (dp[i]) return sums - 2 * i;
        }
        return 0;
    }
};
